File: backend/app/services/message_service.py
"""
Message storage and retrieval.
"""

import uuid

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.db.models import Message, MessageTranslation


async def save_message(
    db: AsyncSession,
    conversation_id: uuid.UUID,
    sender_type: str,
    sender_id: uuid.UUID | None,
    text: str,
    detected_language: str | None = None,
    external_id: str | None = None,
) -> Message:
    """Create and persist a new message."""
    message = Message(
        conversation_id=conversation_id,
        sender_type=sender_type,
        sender_id=sender_id,
        original_text=text,
        detected_language=detected_language,
        external_id=external_id,
    )
    db.add(message)
    await db.flush()
    return message


async def save_translation(
    db: AsyncSession,
    message_id: uuid.UUID,
    source_language: str,
    target_language: str,
    translated_text: str,
    provider: str,
) -> MessageTranslation:
    """Save a translation for a message."""
    translation = MessageTranslation(
        message_id=message_id,
        source_language=source_language,
        target_language=target_language,
        translated_text=translated_text,
        provider=provider,
    )
    db.add(translation)
    await db.flush()
    return translation


async def get_conversation_messages(
    db: AsyncSession,
    conversation_id: uuid.UUID,
) -> list[Message]:
    """Get all messages in a conversation with their translations."""
    result = await db.execute(
        select(Message)
        .options(selectinload(Message.translations))
        .where(Message.conversation_id == conversation_id)
        .order_by(Message.created_at)
    )
    return list(result.scalars().all())


async def get_recent_messages(
    db: AsyncSession,
    conversation_id: uuid.UUID,
    limit: int = 10,
) -> list[Message]:
    """Get the most recent messages (for copilot context)."""
    result = await db.execute(
        select(Message)
        .where(Message.conversation_id == conversation_id)
        .order_by(Message.created_at.desc())
        .limit(limit)
    )
    messages = list(result.scalars().all())
    messages.reverse()  # Chronological order
    return messages

File: backend-ts/src/services/message.service.ts
import { prisma } from "../lib/prisma.js";

export async function saveMessage(params: {
  conversationId: string;
  senderType: string;
  senderId?: string;
  text: string;
  detectedLanguage?: string;
  externalId?: string;
  senderPhone?: string;
  senderName?: string;
  isInternal?: boolean;
}) {
  const message = await prisma.message.create({
    data: {
      conversationId: params.conversationId,
      senderType: params.senderType,
      senderId: params.senderId ?? null,
      originalText: params.text,
      detectedLanguage: params.detectedLanguage ?? null,
      externalId: params.externalId ?? null,
      senderPhone: params.senderPhone ?? null,
      senderName: params.senderName ?? null,
      isInternal: params.isInternal ?? false,
    },
  });

  // Torna a Fila de Chats Orgânica (puxa pro topo ao chegar nova msg)
  await prisma.conversation.update({
    where: { id: params.conversationId },
    data: { updatedAt: new Date() }
  });

  return message;
}

export async function saveAttachment(params: {
  messageId: string;
  type: string;
  mimeType?: string;
  fileName?: string;
  fileSizeBytes?: number;
  sourceUrl?: string;
  providerMediaId?: string;
}) {
  return prisma.messageAttachment.create({
    data: {
      messageId: params.messageId,
      type: params.type,
      mimeType: params.mimeType ?? null,
      fileName: params.fileName ?? null,
      fileSizeBytes: params.fileSizeBytes ?? null,
      sourceUrl: params.sourceUrl ?? null,
      providerMediaId: params.providerMediaId ?? null,
    },
  });
}

export async function saveTranslation(params: {
  messageId: string;
  sourceLanguage: string;
  targetLanguage: string;
  translatedText: string;
  provider: string;
}) {
  return prisma.messageTranslation.create({
    data: {
      messageId: params.messageId,
      sourceLanguage: params.sourceLanguage,
      targetLanguage: params.targetLanguage,
      translatedText: params.translatedText,
      provider: params.provider,
    },
  });
}

export async function getConversationMessages(conversationId: string) {
  const messages = await prisma.message.findMany({
    where: { conversationId, deletedAt: null },
    include: {
      translations: true,
      attachments: true,
      forwardedFrom: { select: { id: true, originalText: true, senderType: true, senderName: true, createdAt: true } },
    },
    orderBy: { createdAt: "asc" },
  });

  // Resolve agent full names for messages sent by agents (senderId is the User.id)
  const agentIds = [
    ...new Set(
      messages
        .filter((m) => m.senderType === "agent" && m.senderId)
        .map((m) => m.senderId!),
    ),
  ];

  if (agentIds.length === 0) return messages;

  const agents = await prisma.user.findMany({
    where: { id: { in: agentIds } },
    select: { id: true, fullName: true },
  });
  const agentNameMap = new Map(agents.map((a) => [a.id, a.fullName]));

  return messages.map((m) => {
    if (m.senderType === "agent" && m.senderId && agentNameMap.has(m.senderId)) {
      return { ...m, senderName: agentNameMap.get(m.senderId) ?? m.senderName };
    }
    return m;
  });
}

export async function getRecentMessages(
  conversationId: string,
  limit = 15,
) {
  return prisma.message.findMany({
    where: { conversationId, deletedAt: null },
    orderBy: { createdAt: "desc" },
    take: limit,
    include: { attachments: true },
  });
}
